ttnn/cpp/ttnn/operations/eltwise/binary_ng/device/binary_ng_device_operation.cpp

```diff
@@ -0,0 +1,232 @@
+
+    const auto& input_shape_a = input_tensor_a.get_logical_shape();
+    const auto input_shape_b =
+        tensor_args.input_tensor_b.has_value() ? tensor_args.input_tensor_b->get_logical_shape() : ttnn::Shape{1, 1};
+
+    constexpr int max_rank = 4;
+    for (int i = 0; i < max_rank; i++) {
+        auto a_dim = i < input_shape_a.rank() ? input_shape_a[-i] : 1;
+        auto b_dim = i < input_shape_b.rank() ? input_shape_b[-i] : 1;
+        TT_FATAL(
+            a_dim == b_dim || a_dim == 1 || b_dim == 1,
+            "ttnn::operations::binary_ng::BinaryNgDeviceOperation: shape mismatch");
+    }
+}
+
+BinaryNgDeviceOperation::spec_return_value_t BinaryNgDeviceOperation::compute_output_specs(
+    const operation_attributes_t& attributes, const tensor_args_t& tensor_args) {
+    const auto& output_tensor = tensor_args.output_tensor;
+    if (output_tensor.has_value()) {
+        return output_tensor->get_tensor_spec();
+    }
+
+    const auto& input_tensor_a = tensor_args.input_tensor_a;
+    const auto input_shape_a = input_tensor_a.logical_shape();
+    const auto& tensor_b = tensor_args.input_tensor_b;
+    const auto input_shape_b = tensor_b.has_value() ? tensor_b->logical_shape() : ttnn::SimpleShape{};
+
+    const int rank_a = input_shape_a.rank();
+    const int rank_b = input_shape_b.rank();
+    const int larger_rank = std::max(rank_a, rank_b);
+
+    // Broadcasting Rules Overview:
+    // - If the two tensors have different ranks, we virtually pad the smaller-rank tensor's shape
+    //   with ones on the left (i.e., higher-order dimensions) until both shapes have the same length.
+    // - For each dimension (starting from the rightmost), the sizes are compatible if:
+    //     - They are equal, or
+    //     - One of them is 1 (the dimension can be broadcast to match the other size).
+    auto compute_broadcasted_output = [rank_a, rank_b, larger_rank](const auto& shape_a, const auto& shape_b) {
+        SmallVector<uint32_t> output_shape(larger_rank, 1);
+        for (int i = -1; i >= -larger_rank; --i) {
+            auto dim_a = (i >= -rank_a) ? shape_a[i] : 1;
+            auto dim_b = (i >= -rank_b) ? shape_b[i] : 1;
+            if (dim_a != 1 && dim_b != 1) {
+                output_shape[i + larger_rank] = dim_a;
+            } else {
+                output_shape[i + larger_rank] = dim_a + dim_b - 1;
+            }
+        }
+        return ttnn::SimpleShape(output_shape);
+    };
+
+    auto output_shape = compute_broadcasted_output(input_shape_a, input_shape_b);
+    return TensorSpec(output_shape, TensorLayout(attributes.dtype, PageConfig(Layout::TILE), attributes.memory_config));
+}
+
+BinaryNgDeviceOperation::program_factory_t BinaryNgDeviceOperation::select_program_factory(
+    const operation_attributes_t&, const tensor_args_t&) {
+    return ProgramFactory{};
+}
+
+BinaryNgDeviceOperation::tensor_return_value_t BinaryNgDeviceOperation::create_output_tensors(
```

Comment:
@TT-BrianLiu @sminakov-tt  is it possible to completely eliminate this method given compute_output_specs is provided and this seem like a default logic?

Reply:
yes default supports optional output tensors

Reply:
so should be able to remove this completely, right?